synapse/replication/tcp/redis.py

```diff
@@ -326,10 +333,47 @@ def __init__(
         )
 
         self.synapse_handler = hs.get_replication_command_handler()
-        self.synapse_stream_name = hs.hostname
+        self.synapse_stream_prefix = hs.hostname
+        self.synapse_subscribed_channels = (
+            RedisDirectTcpReplicationClientFactory.channels_to_subscribe_to_for_config(
+                hs.config
+            )
+        )
 
         self.synapse_outbound_redis_connection = outbound_redis_connection
 
+    @staticmethod
+    def channels_to_subscribe_to_for_config(config: HomeServerConfig) -> List[str]:
+        subscribe_to = []
+
+        if config.worker.run_background_tasks or config.worker.worker_app is None:
+            # If we're the main process or the background worker, we want to process
+            # User IP addresses
+            subscribe_to.append("USER_IP")
+
+        # Subscribe to the following RDATA channels.
+        # We may be able to reduce this in the future.
+        subscribe_to += [
+            "RDATA/account_data",
+            "RDATA/backfill",
+            "RDATA/caches",
+            "RDATA/device_lists",
+            "RDATA/events",
+            "RDATA/federation",
+            "RDATA/groups",
+            "RDATA/presence",
+            "RDATA/presence_federation",
+            "RDATA/push_rules",
+            "RDATA/pushers",
+            "RDATA/receipts",
+            "RDATA/tag_account_data",
+            "RDATA/to_device",
+            "RDATA/typing",
+            "RDATA/user_signature",
+        ]
```

Comment:
I'm not a huge fan of hardcoding the list of streams here. Can we either:

1. Leave `RDATA` on the main channel for now; or
2. Subscribe to `RDATA/*` and use [`PSUBSCRIBE`](https://redis.io/commands/psubscribe/)

Reply:
> Leave RDATA on the main channel for now; or

I think I'd prefer this (mind you, it's still transmitting on the main channel). How do you suggest this work without specifying a list of streams here?
I guess we could have them be subscribed on-demand somewhere else, but I was sort of keen to subscribe them all in one place to be sure we're not sending `REPLICATE` before we are listening on **all** the desired channels.

> Subscribe to RDATA/* and use [PSUBSCRIBE](https://redis.io/commands/psubscribe/)

I am somewhat reluctant to do this because it's not really sorting us out to be able to unsubscribe from select streams in the future (and not to mention that it seems like it's more work for Redis to do pattern matches on each receiver rather than having a concrete set of listeners per channel; not sure how it's implemented in practice though).

Reply:
It sort of depends on what we want the final code to look like. One option that I've been considering is have it so that every handler that wants to listen to a stream has to call something like `ReplicationSubscriber.suIt sort of depends on what we want the final code to look like. One option that I've been considering is have it so that every handler that wants to listen to a stream has to call something like `ReplicationSubscriber.subscribe_to_stream(Stream.NAME, func)` in the handler's `__init__`. Then, when we come to connebscribe_to_stream(Stream.NAME, func)` in the handler's `__init__`. Then, when we come to connect to Redis the `ReplicationSubscriber` has the list of streams the worker is interested in. 

What I really want to avoid is to have to manually list these stream names, its a recipe for it getting out of date and its very non-obvious how it all fits together.

Using `PSUBSCRIBE` would allow us to set up the channels and write to those channels *before* putting all the logic in for doing `ReplicationSubscriber`. OTOH, you could implement the `ReplicationSubscriber` logic before adding the streams

Reply:
I'll give that a go. Otherwise it may be worth cutting out any mention of RDATA in this PR and leaving that rewiring for another PR.

> its very non-obvious how it all fits together.

yeees. That can be said about much of the replication stuff, I don't want to make it much worse though, so will give this a try

Reply:
I've de-scoped the RDATA stuff a bit for now (mostly because I'd appreciate a few thoughts on where exactly to hook it in; the code processing RDATA commands is a bit convoluted and I thought it was going to take a while to untangle), but I've restructured things in a way that I think makes it a bit more approachable to do cleanly.